File: apps/desktop-electron/src/main/ipc.ts
/**
 * DesktopBridge IPC — the renderer <-> platform-shell surface.
 *
 * Audit format (scripts/audit-contract.mjs scans this file): every handler is
 * registered on a LITERAL channel string `bridge:<method>` and the method set
 * is `satisfies DesktopBridge`-checked against @dsh-desktop/protocol, so the
 * registry can never drift from the contract or from the Tauri twin.
 *
 * Push channels are BRIDGE_EVENTS ('dsh:state' / 'dsh:log') — the same
 * constants the preload forwards to the renderer.
 */
import { BrowserWindow, app, ipcMain, shell } from 'electron';
import {
  BRIDGE_EVENTS,
  DESKTOP_BRIDGE_METHODS,
  type DesktopBridge,
  type DesktopSettings,
  type HostEndpoint,
  type HostStatus,
  type InstalledPlugin,
  type Profile,
} from '@dsh-desktop/protocol';
import type { DshSidecar } from './sidecar';
import type { Launcher } from './launcher';
import type { ShellGeneration } from './shell';
import type { SettingsStore } from './settings-store';

export interface DesktopBridgeDeps {
  sidecar: DshSidecar;
  generation: ShellGeneration;
  launcher: Launcher;
  openSettings: () => void;
  userDataDir: string;
  settingsStore: SettingsStore;
  /** Read-only plugin inventory of the current profile (dsh.bundle/client manifests). */
  pluginList: () => InstalledPlugin[];
  /** Collect + write the diagnostics report; returns the written file path. */
  exportDiagnostics: () => { path: string };
  log: (level: 'info' | 'warn' | 'error', line: string) => void;
  /** Subscribe to (already masked) main-process log lines for the 'dsh:log' push channel. */
  logOnLine: (subscriber: (entry: { level: 'info' | 'warn' | 'error'; line: string }) => void) => () => void;
  /** Manual host_start re-arms the crash supervisor after a give-up. */
  onHostStart?: () => void;
}

/** Bridge surface implementation, type-checked against the canonical contract. */
function createBridge(deps: DesktopBridgeDeps): DesktopBridge {
  return {
    host_start: (): Promise<HostEndpoint> => {
      deps.onHostStart?.();
      return deps.sidecar.spawn();
    },
    host_stop: async (): Promise<void> => {
      await deps.sidecar.stop();
    },
    host_restart: (): Promise<HostEndpoint> => deps.sidecar.restart(),
    host_status: async (): Promise<HostStatus> => deps.sidecar.status,

    window_show: async (): Promise<void> => {
      deps.generation.show();
    },
    window_hide: async (): Promise<void> => {
      deps.generation.hide();
    },
    window_focus: async (): Promise<void> => {
      deps.generation.focus();
    },
    window_open_settings: async (): Promise<void> => {
      deps.openSettings();
    },

    profile_list: async (): Promise<Profile[]> => deps.launcher.listProfiles(),
    profile_current: async (): Promise<Profile> => deps.launcher.getCurrentProfile(),
    profile_switch: async (name: string): Promise<void> => {
      await deps.launcher.switchProfile(name);
    },

    open_data_dir: async (): Promise<void> => {
      await shell.openPath(deps.userDataDir);
    },
    open_external: async (url: string): Promise<void> => {
      // Fence: only these schemes may leave the app; everything else is a
      // renderer bug or an injection attempt.
      let parsed: URL;
      try {
        parsed = new URL(url);
      } catch {
        throw new Error(`open_external: not a valid URL: ${url}`);
      }
      if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:' && parsed.protocol !== 'mailto:') {
        throw new Error(`open_external: scheme not allowed: ${parsed.protocol}`);
      }
      await shell.openExternal(parsed.href);
    },
    get_app_version: async (): Promise<string> => app.getVersion(),

    settings_get: async (): Promise<DesktopSettings> => deps.settingsStore.get(),
    settings_set: async (patch: Partial<DesktopSettings>): Promise<DesktopSettings> => {
      const next = deps.settingsStore.set(patch);
      // Zoom applies to the live webContents; the other keys are read at
      // window close/mount time.
      deps.generation.setZoom(next.zoomFactor);
      return next;
    },
    plugin_list: async (): Promise<InstalledPlugin[]> => deps.pluginList(),
    diagnostics_export: async (): Promise<{ path: string }> => deps.exportDiagnostics(),
  };
}

export interface BridgeIpcRegistration {
  unregister: () => void;
}

export function registerDesktopBridgeIpc(deps: DesktopBridgeDeps): BridgeIpcRegistration {
  const bridge = createBridge(deps);

  // Literal channel names — the contract audit greps for exactly these.
  ipcMain.handle('bridge:host_start', () => bridge.host_start());
  ipcMain.handle('bridge:host_stop', () => bridge.host_stop());
  ipcMain.handle('bridge:host_restart', () => bridge.host_restart());
  ipcMain.handle('bridge:host_status', () => bridge.host_status());
  ipcMain.handle('bridge:window_show', () => bridge.window_show());
  ipcMain.handle('bridge:window_hide', () => bridge.window_hide());
  ipcMain.handle('bridge:window_focus', () => bridge.window_focus());
  ipcMain.handle('bridge:window_open_settings', () => bridge.window_open_settings());
  ipcMain.handle('bridge:profile_list', () => bridge.profile_list());
  ipcMain.handle('bridge:profile_current', () => bridge.profile_current());
  ipcMain.handle('bridge:profile_switch', (_event, name: string) => bridge.profile_switch(name));
  ipcMain.handle('bridge:open_data_dir', () => bridge.open_data_dir());
  ipcMain.handle('bridge:open_external', (_event, url: string) => bridge.open_external(url));
  ipcMain.handle('bridge:get_app_version', () => bridge.get_app_version());
  ipcMain.handle('bridge:settings_get', () => bridge.settings_get());
  ipcMain.handle('bridge:settings_set', (_event, patch: Partial<DesktopSettings>) => bridge.settings_set(patch));
  ipcMain.handle('bridge:plugin_list', () => bridge.plugin_list());
  ipcMain.handle('bridge:diagnostics_export', () => bridge.diagnostics_export());

  const sendToAll = (channel: string, payload: unknown): void => {
    for (const win of BrowserWindow.getAllWindows()) {
      if (!win.isDestroyed()) win.webContents.send(channel, payload);
    }
  };

  const onSidecarState = (status: HostStatus): void => {
    sendToAll(BRIDGE_EVENTS.state, status);
  };
  const onLogLine = (entry: { level: 'info' | 'warn' | 'error'; line: string }): void => {
    sendToAll(BRIDGE_EVENTS.log, entry);
  };

  deps.sidecar.on('state', onSidecarState);
  const unsubscribeLog = deps.logOnLine(onLogLine);

  return {
    unregister: () => {
      deps.sidecar.off('state', onSidecarState);
      unsubscribeLog();
      for (const method of DESKTOP_BRIDGE_METHODS) {
        ipcMain.removeHandler(`bridge:${method}`);
      }
    },
  };
}

File: apps/desktop-tauri/src-tauri/src/ipc.rs
//! DesktopBridge IPC commands — the Tauri half of packages/protocol/src/
//! bridge.ts. The 14 command names, parameters and result types must match
//! `DESKTOP_BRIDGE_METHODS` verbatim; scripts/audit-contract.mjs diffs this
//! file's `#[tauri::command]` set against the protocol list.

use tauri::{AppHandle, Manager};
use tauri_plugin_opener::OpenerExt;

use crate::diagnostics;
use crate::launcher::{self, Profile};
use crate::plugin_inventory;
use crate::settings;
use crate::settings_window;
use crate::sidecar::{AppState, HostEndpoint, HostStatus};

// -- 生命周期 ----------------------------------------------------------------

/// Spawn the sidecar and wait for the ready line.
/// Idempotent: returns the live endpoint when already running.
#[tauri::command]
pub async fn host_start(state: tauri::State<'_, AppState>) -> Result<HostEndpoint, String> {
    state.sidecar.start().await
}

/// Kill the sidecar process tree.
#[tauri::command]
pub async fn host_stop(state: tauri::State<'_, AppState>) -> Result<(), String> {
    state.sidecar.stop().await
}

/// Stop, then spawn again; resolves with the fresh endpoint.
#[tauri::command]
pub async fn host_restart(state: tauri::State<'_, AppState>) -> Result<HostEndpoint, String> {
    state.sidecar.restart().await
}

/// Current status snapshot (no waiting).
#[tauri::command]
pub fn host_status(state: tauri::State<'_, AppState>) -> Result<HostStatus, String> {
    Ok(state.sidecar.status())
}

// -- 窗口 --------------------------------------------------------------------

fn main_window(app: &AppHandle) -> Result<tauri::WebviewWindow, String> {
    app.get_webview_window("main")
        .ok_or_else(|| "main window is not available".to_string())
}

#[tauri::command]
pub fn window_show(app: AppHandle) -> Result<(), String> {
    let window = main_window(&app)?;
    let _ = window.show();
    let _ = window.unminimize();
    Ok(())
}

#[tauri::command]
pub fn window_hide(app: AppHandle) -> Result<(), String> {
    let window = main_window(&app)?;
    let _ = window.hide();
    Ok(())
}

#[tauri::command]
pub fn window_focus(app: AppHandle) -> Result<(), String> {
    let window = main_window(&app)?;
    let _ = window.unminimize();
    let _ = window.set_focus();
    Ok(())
}

/// Open (or focus) the dedicated settings window.
#[tauri::command]
pub fn window_open_settings(app: AppHandle) -> Result<(), String> {
    settings_window::open_settings(&app)
}

// -- Profile ------------------------------------------------------------------

#[tauri::command]
pub fn profile_list() -> Result<Vec<Profile>, String> {
    Ok(launcher::list_profiles())
}

#[tauri::command]
pub fn profile_current() -> Result<Profile, String> {
    Ok(launcher::profile_detail(&launcher::current_profile()))
}

/// Persist the selection and apply it immediately: a running host restarts on
/// the new profile (`--profile` is a launcher flag, fixed at spawn time).
#[tauri::command]
pub async fn profile_switch(
    name: String,
    state: tauri::State<'_, AppState>,
) -> Result<(), String> {
    launcher::switch_profile(&name).map_err(|err| err.to_string())?;
    state.sidecar.restart_if_running().await
}

// -- 系统 ---------------------------------------------------------------------

/// Shared with the tray menu (ipc::open_data_dir_impl).
pub(crate) fn open_data_dir_impl(app: &AppHandle) -> Result<(), String> {
    let dir = launcher::dsh_home();
    std::fs::create_dir_all(&dir).map_err(|err| format!("cannot create data dir: {err}"))?;
    app.opener()
        .open_path(dir.to_string_lossy().to_string(), None::<&str>)
        .map_err(|err| format!("failed to open data dir: {err}"))
}

#[tauri::command]
pub fn open_data_dir(app: AppHandle) -> Result<(), String> {
    open_data_dir_impl(&app)
}

/// Open a URL in the system browser / mail client. Only http, https and
/// mailto are allowed — everything else (file:, javascript:, ...) is refused.
#[tauri::command]
pub fn open_external(app: AppHandle, url: String) -> Result<(), String> {
    let parsed = tauri::Url::parse(&url).map_err(|_| format!("invalid url: {url}"))?;
    match parsed.scheme() {
        "http" | "https" | "mailto" => {}
        _ => {
            return Err(format!(
                "only http, https and mailto urls may be opened externally (got {})",
                parsed.scheme()
            ))
        }
    }
    app.opener()
        .open_url(url, None::<&str>)
        .map_err(|err| format!("failed to open external url: {err}"))
}

#[tauri::command]
pub fn get_app_version() -> Result<String, String> {
    Ok(env!("CARGO_PKG_VERSION").to_string())
}

// -- v1.1：设置 / 插件清单 / 诊断 ------------------------------------------

/// Load shell settings (defaults when the file is missing or corrupt).
#[tauri::command]
pub fn settings_get() -> Result<settings::DesktopSettings, String> {
    Ok(settings::load_from(&settings::settings_path()))
}

/// Validate → persist → apply. Zoom lands on the live webContents immediately;
/// close/start keys are read at close/mount time.
#[tauri::command]
pub fn settings_set(patch: serde_json::Value) -> Result<settings::DesktopSettings, String> {
    let current = settings::load_from(&settings::settings_path());
    let next = settings::apply_patch(&current, &patch).map_err(|err| err.0)?;
    settings::save_to(&settings::settings_path(), &next).map_err(|err| err.0)?;
    // TODO(verify-with-cargo): WebviewWindow::set_zoom exists in tauri 2.x but
    // is not exercised by the local reference crate — CI compiles this.
    if let Some(app) = crate::APP.get() {
        if let Some(window) = app.get_webview_window("main") {
            let _ = window.set_zoom(next.zoom_factor);
        }
    }
    Ok(next)
}

/// Read-only inventory of DSH plugins in the current profile.
#[tauri::command]
pub fn plugin_list() -> Result<Vec<plugin_inventory::InstalledPlugin>, String> {
    let profile = launcher::current_profile();
    Ok(plugin_inventory::list_profile_plugins(&launcher::profiles_dir().join(profile)))
}

/// Collect + write the diagnostics report to the desktop dir, reveal it.
#[tauri::command]
pub fn diagnostics_export(app: AppHandle, state: tauri::State<'_, AppState>) -> Result<serde_json::Value, String> {
    let host_status = serde_json::to_value(state.shared.status()).map_err(|err| err.to_string())?;
    let host_hello = state
        .shared
        .hello_info()
        .and_then(|hello| serde_json::to_value(hello).ok())
        .unwrap_or(serde_json::Value::Null);
    let profiles = serde_json::to_value(launcher::list_profiles()).map_err(|err| err.to_string())?;
    let current = settings::load_from(&settings::settings_path());
    let settings_value = serde_json::to_value(current).map_err(|err| err.to_string())?;
    let desktop_state = std::fs::read_to_string(launcher::desktop_state_path())
        .ok()
        .and_then(|text| serde_json::from_str::<serde_json::Value>(&text).ok())
        .unwrap_or(serde_json::Value::Null);

    let logs_dir = app.path().app_data_dir().ok().map(|dir| dir.join("logs"));
    let dsh_home = launcher::dsh_home();
    let input = diagnostics::DiagnosticsInput {
        app_version: env!("CARGO_PKG_VERSION"),
        platform: std::env::consts::OS,
        dsh_home: &dsh_home.to_string_lossy(),
        logs_dir: logs_dir.as_deref().unwrap_or(std::path::Path::new(".")),
        host_status,
        host_hello,
        profiles,
        settings: settings_value,
        desktop_state,
    };
    let report = diagnostics::collect(&input);
    let dir = dirs::desktop_dir().or_else(dirs::data_dir).unwrap_or_else(std::env::temp_dir);
    let file = diagnostics::write_report(&dir, &report)?;
    Ok(serde_json::json!({ "path": file.to_string_lossy() }))
}
